### PR TITLE
Add support for smart contracts and set up chain_extension

### LINKED ISSUE #34: extend smart contract AMM to full asset support

### PR DESCRIPTION
**Pull request by ebma**
- [x] Add `pallet-contracts` pallet and configure it for use with runtime
- [x] Implemet chain extension that offers functions for smart contracts to interact with the runtime
- [x] Increase the initial balance of accounts to 1MUnit (A higher initial balance is needed to make the deployment of smart contracts feasible because the `endowment` should be at least 1000 Units)

I don't know why the old commits are also shown here, maybe because I cherry-picked 24bc867 onto the branch.

Part of #34.

### CODE REVIEW

**Comment by morgueye4** [comment]
LGTM

**Comment by ebma** [comment]
Not yet ready for merge because fetching the balances is not working properly.

**Comment by ebma** [comment]
The balance issue is fixed so now this can be merged after review.

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -424,22 +422,15 @@ impl ChainExtension<Runtime> for BalanceChainExtension {
                         let mut asset_code_array: [u8; 12] = Default::default();
                         asset_code_array.copy_from_slice(&input[64..]);
                         let asset_str: &str = str::from_utf8(&asset_code_array).unwrap();
```
Comment (gonzamontiel) [inline runtime/src/lib.rs]: It's better if we map the error so it doesn't just panic. Same below.

Comment (gonzamontiel) [reply runtime/src/lib.rs]: Already fixed that.

**Comment by gonzamontiel** [comment]
I reviewed the branch, added a comment, fixed it. looks ok to me for merging. @morgueye4 please merge this asap so we can deploy the latest version tomorrow.

**Pull request merged**